build_barback/lib/src/transformer/crawl_imports.dart

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import 'dart:async';
+
+import 'package:analyzer/analyzer.dart' show parseDirectives;
+import 'package:analyzer/src/generated/engine.dart' show TimestampedData;
+import 'package:analyzer/src/generated/source.dart';
+import 'package:build/build.dart';
+import 'package:path/path.dart' as p;
+
+import 'package:analyzer/dart/ast/ast.dart';
+
+/// Crawl Dart files outside of the SDK transitively imported from any library
+/// in [entryPoints] and cache it as a [Source].
+Future<Map<Uri, Source>> crawlImports(
+    Iterable<AssetId> entryPoints, Future<String> read(AssetId id)) async {
+  var roots = entryPoints.where((id) => id.path.endsWith('.dart'));
+  var assets = await _crawlAsync(roots, read, _assetImports);
+  var results = <Uri, Source>{};
+  for (var asset in assets.keys) {
+    results[asset.uri] = _toSource(asset, assets[asset]);
+  }
+  return results;
+}
+
+Iterable<AssetId> _assetImports(AssetId id, String source) {
+  var unit = parseDirectives(source);
+  return unit.directives
+      .where((d) => d is UriBasedDirective)
+      .map((d) => d as UriBasedDirective)
+      .map((d) => d.uri.stringValue)
```

Comment:
nit: I would combine these and just do `(d as UriBasedDirective).uri.stringValue`